test/ecdhtest.c

```diff
@@ -462,6 +461,105 @@ static int ecdh_kat(BIO *out, const ecdh_kat_t *kat)
     return rv;
 }
 
+#include "ecdhtest_cavs.h"
+
+static int ecdh_cavs_kat(BIO *out, const ecdh_cavs_kat_t *kat)
+{
+    int rv = 0, is_char_two = 0;
+    EC_KEY *key1 = NULL;
+    EC_POINT *pub = NULL;
+    const EC_GROUP *group = NULL;
+    BIGNUM *bnz = NULL, *x = NULL, *y = NULL;
+    unsigned char *Ztmp = NULL, *Z = NULL;
+    size_t Ztmplen, Zlen;
+    BIO_puts(out, "Testing ECC CDH Primitive SP800-56A with ");
+    BIO_puts(out, OBJ_nid2sn(kat->nid));
+
+    /* dIUT is IUT's private key */
+    if ((key1 = mk_eckey(kat->nid, kat->dIUT)) == NULL)
+        goto err;
+    /* these are cofactor ECDH KATs */
+    EC_KEY_set_flags(key1, EC_FLAG_COFACTOR_ECDH);
+
+    if ((group = EC_KEY_get0_group(key1)) == NULL)
+        goto err;
+    if ((pub = EC_POINT_new(group)) == NULL)
+        goto err;
+
+    if (EC_METHOD_get_field_type(EC_GROUP_method_of(group)) == NID_X9_62_characteristic_two_field)
+        is_char_two = 1;
+
+    /* (QIUTx, QIUTy) is IUT's public key */
+    if(!BN_hex2bn(&x, kat->QIUTx))
+        goto err;
+    if(!BN_hex2bn(&y, kat->QIUTy))
+        goto err;
+    if (is_char_two) {
+#ifndef OPENSSL_NO_EC2M
+        if (!EC_POINT_set_affine_coordinates_GF2m(group, pub, x, y, NULL))
+#endif
+            goto err;
```

Comment:
Thanks for looking.

Hmm I don't see any easy way to test if it's a binary curve from just the NID. EC_GROUP_new_by_curve_name will fail before I can even fetch the field type.

Currently, this is handled in ecdhtest_cavs.h with ifdef statements, so the binary tests only get compiled in if binary curves are enabled.

Is this sufficient or do you want something more elaborate? BTW tested with both no-ec2m on and off.